ENH, WIP: Add 2-D fitting function for polynomials

AWehrhahn opened the pull request:
ENH: add 2d polynomial fit function to numpy.polynomial.

charris commented:
New functions need to be discussed on the mailing list.

In practice, I've found it useful to have a way to limit which coefficients are allowed to vary. The case where `deg(x) + deg(y) <= N` for some N turns up fairly often. We also like to keep functions common between all the polynomial types when possible, look at how the current 1-D fitting is implemented.

The style changes should be in a separate PR and aren't required by the current NumPy standard, spaces around `*` and `/` aren't needed.

charris commented:
If you are feeling more ambitious, it would be nice to design a 2-D class to complement the current 1-class. It would have partial derivative methods, maybe integration too. It will not be a trivial effort, I believe SciPy (@rgommers) is also looking into multidimensional methods for interpolation and this might be related.

AWehrhahn commented:
Where can I find the mailing list, I just followed the guide and it didn't mention anything.

I included a case for `deg(x) + deg(y) <= N` with the max_degree parameter.
What do you mean with keeping functions common between polynomial types? polyfit1d is implemented more or less the same way?

The style changes where unintentional, that was just the auto-formatter doing its work. Will reverse them in the next commit.

I don't think I will have time to do a proper 2d polynomial class, maybe at some later time

AWehrhahn committed:
Polyfit2d initial. Now without unnecessary style changes

charris commented:
> What do you mean with keeping functions common between polynomial types?

That all six polynomial types have the same basic set of functions. See `numpy/polynomial/polyutils.py` for an example of abstracting the common fit functionality in the `_fit` function. That function is then used to implement 1-D fitting for all the polynomial types.

The mailing list is  numpy-discussion@python.org, but if you initiate a discussion it is best to subscribe at https://mail.python.org/mailman/listinfo/numpy-discussion.

AWehrhahn committed:
Moved 2d fit into polyutils, similar to the 1d fit

AWehrhahn committed:
bugfixes, so that test work

grlee77 commented:
I am not an expert in this area, but have wanted to do this previously (but in 3D rather than 2D) and some searching turned up scikit-learn as an existing solution. It is possible to fit n-dimensional polynomials using scikit-learn's `PolynomialFeatures` combined with a linear model like `LinearRegression` or  `Ridge`. The [examples given in the docs](
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html#examples-using-sklearn-preprocessing-polynomialfeatures
) are for 1D, but the underlying `PolynomialFeatures` can take multiple coordinates for 2D or higher dimensional data.

A benefit of the API of `polyfit2d` proposed here, is that it should be easier to quickly understand for a user just wanting to do 2D interpolation without the overhead of having to learn about scikit-learn classes / pipelines.

AWehrhahn committed:
2d fits for other polynomial classes

AWehrhahn commented:
I moved the 2d fit procedure into polyutils and added functions in the other polynomial classes for the 2d fit. I had to disable the scaling though as the shift works differently (I assume).

charris commented on numpy/polynomial/polynomial.py:
```diff
@@ -1336,6 +1337,89 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None):
+        If given the maximum combined degree of the coefficients is limited
+        to this value, i.e. all terms with `n` + `m` > max_degree are set to 0.
+        The default is None.
+    scale : bool, optional
```
Scaling is even more important for the other polynomial basis whose weights have compact support, which is why I'd like to see a class for 2-d objects at some point so that the extra information, also including the selection of coefficients, could be carried along. I don't think it belongs in the base fitting function, however. It is also possible to scale the 2d domain to `[-1, 1]` in each direction and get good results in most cases, images and such.

charris commented:
Wow, nice job, just a couple of comments.

- Code lines need to be < 80 characters in length, a few exceed that.
- Docstring lines should be <= 75 characters in length.
- There are a couple of typos, it would be good to run a spellcheck on everything.
- Scale is only present in the power basis, and should probably be omitted on that count.

I'd really like to figure out a way to reduce the amount of repetition in the docstrings, but haven't solved that problem yet. The polynomial documentation is a significant part of the total NumPy documentation :) There is an argument to be made for the polynomials to be their own project.

If you have any ideas of how to make a 2d class, I'd like to hear it. Such a class would include partial derivatives, along with scaling and offsets.

I'm wondering if it would make sense to use a mask for selecting the fitted coefficients or would that be too general? The current setup covers the most common options.

charris commented:
I still need to review the code in detail, the tests in particular.

eric-wieser commented:
I have some ideas on ND versions, possibly with mixed polynomial types - will try to dump them in a hackmd document or issue at some point.

eric-wieser commented on numpy/polynomial/polyutils.py:
```diff
@@ -690,6 +690,280 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
+    idx : array of shape (n * m, 2)
+        pairs of indices to the coefficient matrix
+    """
+    idx = np.indices(shape)
```
Can you add `assert len(shape) == 2` here?

AWehrhahn replied:
Done

AWehrhahn committed:
Add assertion of shape for the 2d index

AWehrhahn committed:
Fix long lines

AWehrhahn committed:
Fix some typos

AWehrhahn commented:
Adding a mask for the selection of the fitted polynomial degrees would be pretty simple, considering that is what i do for max_degree anyway.

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,129 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    --------
+
+    """
+    return pu._fit2d(chebvander2d, x, y, z, deg, rcond, full, w, max_degree, False)
```
Tempting to pass `chebvander` in directly here, and let `pu._fit2d` call `pu._vander2d`

AWehrhahn replied:
I mean that is exactly what chebvander2d is doing. No need to repeat that here I think.

eric-wieser replied:
The reason I suggest this is that you if you passed `chebvander` directly onto `_vander_nd`, then that would allow fitting of mixed-axis polynomicals, eg `pu._fit2d([chebvander, polyvander], ...)`

AWehrhahn replied:
How about:

    if not callable(vander2d_f):
        vander2d_f = lambda x, y, deg: _vander_nd_flat(vander2d_f, (x, y), deg)

I.e. allow both?

AWehrhahn committed:
Polyfit with degree mask
Remove shifting of the polynomial
Now scales to the maximum of the input values

AWehrhahn commented:
I added the degree handling, using a mask. One can then pass a 2d array with the same size as the output coefficient matrix, where degrees that are to be used are set to 1 (or more), and the others set to 0.

eric-wieser commented on numpy/polynomial/polyutils.py:
```diff
@@ -684,6 +684,264 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
+    idx = _get_coeff_idx(coeff.shape)
+    for i, j in idx:
+        coeff[i, j] /= scale_x ** i * scale_y ** j
+    return coeff
```
I think this can be vectorized, will come back to it...

AWehrhahn replied:
Done

eric-wieser commented on numpy/polynomial/polynomial.py:
```diff
@@ -1336,6 +1337,89 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None):
+        to this value, i.e. all terms with `n` + `m` > max_degree are set to 0.
+        The default is None.
+    scale : bool, optional
+        Wether to scale the input arrays x and y to mean 0 and variance 1, to avoid numerical overflows.
```
```suggestion
        Whether to scale the input arrays x and y to mean 0 and variance 1, to avoid numerical overflows.
```

AWehrhahn committed:
Update numpy/polynomial/polynomial.py

Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,129 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    least squares fit to the data values `y` given at points `x`.
+    The fitted polynomial(s) are in the form
+
+    .. math::  p(x, y) = c_00 + c_10 * T_10(x, y) + ... + c_nm * T_nm(x, y),
```
You need to write this as `T_{10}` etc, else it renders as 
![image](https://user-images.githubusercontent.com/425260/67670034-f2fc9580-f96a-11e9-93f3-072f64a30c4c.png)

As an aside - is it the case that `T_{10}(x, y) = T_1(x)T_0(y)`? I think this would be clearer, since we don't defined a two-argument `T` anyway.

AWehrhahn replied:
I have no idea about Chebyshev polynomials, but that sounds reasonbable. So I changed it

AWehrhahn replied:
I also changed it in the other polynomial types

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,129 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+        weights are chosen so that the errors of the products ``w[i]*y[i]``
+        all have the same variance.  The default value is None.
+
+        .. versionadded:: 1.5.0
```
```suggestion
```

This function didn't even exist in 1.5.0!

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,129 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+
+    .. math::  p(x, y) = c_00 + c_10 * T_10(x, y) + ... + c_nm * T_nm(x, y),
+
+    where `n` and `m` are `deg`.
```
Perhaps a good idea to add `..versionadded:: 1.19.0` here (I doubt very much this will make 18)

AWehrhahn replied:
Added

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,129 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
 
+def chebfit2d(x, y, z, deg, rcond=None, full=False, w=None, max_degree=None):
+    """
+    2d Least squares fit of Chebyshev series to data.
```
I wonder if we should construct this docstring from a template, it irritates me to see such a large docstring for such a small function body.

AWehrhahn committed:
Fix docstrings

AWehrhahn committed:
vectorize coefficient scaling

AWehrhahn committed:
Fixed docstring again

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,130 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    2d Least squares fit of Chebyshev series to data.
+
+    Return the coefficients of a Chebyshev series of degree `deg` that is the
+    least squares fit to the data values `y` given at points `x`.
```
```suggestion
    least squares fit to the data values `z` given at points `(x, y)`.
```

eric-wieser replied:
This is why this repeated docstring is so dangerous, it almost ensures that one or many will be wrong.

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1647,6 +1647,130 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    least squares fit to the data values `y` given at points `x`.
+    The fitted polynomial(s) are in the form
+
+    .. math::  p(x, y) = c_{00} + c_{10} * T_1(x) T_0(y) + ... + c_{nm} * T_n(x) T_m(y),
```
```suggestion
    .. math::  p(x, y) = c_{00} + c_{10} T_1(x) T_0(y) + ... + c_{nm} T_n(x) T_m(y),
```

charris commented:
Going to push this off to 1.20. It needs a rebase. We may try for a generalize nd fit.

AWehrhahn committed:
polyfitnd

AWehrhahn committed:
Implement N dimensional fit

AWehrhahn commented:
I rebased the version to the numpy repository. And changed it to a ND fit, instead of just 2D.
For some reason my local numpy version did not compile, so I couldn't run the tests, but some simple tests look good.

AWehrhahn commented on numpy/polynomial/polyutils.py:
```diff
@@ -725,6 +725,179 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
+    lhs = vandernd_f2(coords, deg)
+    rhs = data
+
+    # Determine the positions within the vandermode matrix
+    # TODO: how does that work with non standard polynomials?
+    idx = np.zeros((lhs.shape[1], ndim), dtype=int)
+    for i in range(ndim):
+        c = np.ones(ndim)
+        c[i] = 2
+        factors = vandernd_f2(c, deg) // 2
+        idx[:, i] = factors[0]
```
I am not very happy with this part, and I don't really know how this will work with not "polynomial" polynomials (e.g. hermite polynomials).

AWehrhahn committed:
Fix the fit2d function in the other polynomial types

AWehrhahn committed:
Fix initial checks

AWehrhahn committed:
Fix the vandermode matrix indices
Fix some of the documentation
Fix the tests by using meshgrid
1D case should be consistent with _fit

AWehrhahn reviewed (commented):
I fixed this in a later version

eric-wieser commented on numpy/polynomial/polyutils.py:
```diff
@@ -725,6 +726,182 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
     else:
         return c
 
+def _fitnd(vandernd_f, coords, data, deg=1, rcond=None, full=False, w=None,
```
`w` is not documented

eric-wieser commented on numpy/polynomial/polyutils.py:
```diff
@@ -725,6 +726,182 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
+            raise TypeError(f"expected non-empty vector for coords[{i}]")
+        if len(x) != npoints:
+            raise TypeError(f"expected coords[{i}] and data to have the same length")
+    if w is not None:
```
This is always `True` due to line 774. Did you mean to put line 774 after this point?

AWehrhahn replied:
line 774 is only called if w is not None ?

eric-wieser replied:
Ah, my mistake.

eric-wieser commented on numpy/polynomial/polyutils.py:
```diff
@@ -725,6 +726,182 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
+
+    Parameters
+    ----------
+    vander2d_f : {function(array_like, ..., int) -> ndarray, list of function(array_like, int) -> ndarray}
+        The 2d vander function, such as ``polyvander2d``,
+        or a list of 1d vander functions for each dimension
```
Can we remove the complexity here and only allow the list of 1d vander functions?

AWehrhahn replied:
Sure, will change that soon

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1663,6 +1663,130 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
     return pu._fit(chebvander, x, y, deg, rcond, full, w)
 
 
+def chebfit2d(x, y, z, deg, rcond=None, full=False, w=None, max_degree=None):
```
Any reason not to immediately go to nd?
```suggestion
def chebfitnd(xs, fxs, deg, *, rcond=None, full=False, w=None, max_degree=None):
```

AWehrhahn replied:
The function call is slightly different. The 2d version has seperate paramaters for each dimension x and y. While the nd version only has 1 parameter (x, y) with all coordinates in a tuple.
The 2d version is more consistent with the existing functions for e.g polyval2d.

eric-wieser replied:
So, I'd argue that in hindsight the 2d functions were a mistake - we should have gone straight from 1d to nd, rather than blowing up our API with intermediate integers.

Since what you're adding is a new API, what I'm arguing is that we should learn from our mistake, and go straight to writing the nd function.

eric-wieser replied:
If you feel strongly about the calling convention, you could use instead:
```
def chebfitnd(*args, *, deg, rcond=None, full=False, w=None, max_degree=None):
    *xs, fxs = args
```

AWehrhahn replied:
I have no particular attachement to the calling convention

pbrod replied:
Why not deprecating all the XXX2d and XXX3d functions (like chebfit2d, chebfit3d, chebval2d, chebval3d, chebgrid2d,    chebgrid3d, chebvander2d,   chebvander3d, etc.) since they all are obsolete once you have the XXXnd functions (like  chebfitnd,  chebvalnd, chebgridnd, chebvandernd, ...)?

eric-wieser commented on numpy/polynomial/chebyshev.py:
```diff
@@ -1663,6 +1663,130 @@ def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    --------
+
+    """
+    return pu._fitnd(chebvander2d, (x, y), z, deg, rcond, full, w, max_degree)
```
With my above suggestion,

```suggestion
    return pu._fitnd([chebvander] * len(xs), xs, fxs, deg, rcond, full, w, max_degree)
```

AWehrhahn committed:
Fix tests
Fix documentation

AWehrhahn committed:
Fix typo

AWehrhahn committed:
Bugfix
